msal/src/main/java/com/microsoft/identity/client/AuthenticationActivity.java

```diff
@@ -125,29 +131,70 @@ private void warmUpCustomTabs() {
                 mCustomTabsServiceConnection
         );
 
+        boolean initCustomTabsWithSession = true;
+        try {
+            // await returns true if count is 0, false if action times out
+            // invert this boolean to indicate if we should skip warming up
+            boolean timedOut = !latch.await(CUSTOMTABS_MAX_CONNECTION_TIMEOUT, TimeUnit.SECONDS);
+            if (timedOut) {
+                // if the request timed out, we don't actually know whether or not the service connected.
+                // to be safe, we'll skip warmup and rely on mCustomTabsServiceIsBound
+                // to unbind the Service when onStop() is called.
+                initCustomTabsWithSession = false;
+                Logger.warning(TAG, null, "Connection to CustomTabs timed out. Skipping warmup.");
+            }
+        } catch (InterruptedException e) {
+            Logger.error(TAG, null, "Failed to connect to CustomTabs. Skipping warmup.", e);
+            initCustomTabsWithSession = false;
+        }
+
+        final CustomTabsIntent.Builder builder = initCustomTabsWithSession ? new CustomTabsIntent.Builder(mCustomTabsSession) : new CustomTabsIntent.Builder();
+
         // Create the Intent used to launch the Url
-        mCustomTabsIntent = new CustomTabsIntent.Builder(mCustomTabsSession)
-                .setShowTitle(true)
-                .build();
+        mCustomTabsIntent = builder.setShowTitle(true).build();
         mCustomTabsIntent.intent.setPackage(mChromePackageWithCustomTabSupport);
     }
 
-    @NonNull
-    private CustomTabsServiceConnection createCustomTabsServiceConnection() {
-        return new CustomTabsServiceConnection() {
-            @Override
-            public void onCustomTabsServiceConnected(ComponentName name, CustomTabsClient client) {
-                mCustomTabsServiceIsBound = true;
-                mCustomTabsClient = client;
-                mCustomTabsClient.warmup(0L);
-                mCustomTabsSession = mCustomTabsClient.newSession(null);
+    private static class MsalCustomTabsServiceConnection extends CustomTabsServiceConnection {
+
+        private final WeakReference<AuthenticationActivity> mAuthenticationActivityWeakReference;
```

Comment:
Like I mentioned above, having a reference to enclosing class is a bad sign. If it won't cause too much trouble, i would suggest we have the customsTabSession and customTabsClient declared inside of the connection, once it's connected you'll set those values. We can do connection.getTabSesson or connection.getTabsClient for it's enclosing class to use.